lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,119 @@
+"""
+Middleware for Mobile APIs
+"""
+from datetime import datetime
+from django.core.cache import cache
+from django.http import HttpResponse
+from pytz import UTC
+from mobile_api.mobile_platform import MobilePlatform
+from mobile_api.models import AppVersionConfig
+from mobile_api.utils import parsed_version
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+from request_cache.middleware import RequestCache
+
+
+class AppVersionUpgrade(object):
+    """
+    Middleware class to keep track of mobile application version being used
+    """
+    LATEST_VERSION_HEADER = 'EDX-APP-LATEST-VERSION'
+    UPGRADE_DEADLINE_HEADER = 'EDX-APP-UPGRADE-DATE'
+    NO_LAST_SUPPORTED_DATE = 'NO_LAST_SUPPORTED_DATE'
+    NO_LATEST_VERSION = 'NO_LATEST_VERSION'
+
+    def process_request(self, request):
+        """
+        raises HTTP Upgrade Require error if request is from mobile native app and
+        user app version is no longer supported
+        """
+        user_agent = request.META.get('HTTP_USER_AGENT')
+        request_cache_dict = self.set_request_cache(request, user_agent)
+        last_supported_date = (request_cache_dict[self.UPGRADE_DEADLINE_HEADER]
```

Comment:
This can be shortened to `last_supported_date = request_cache_dict.get(self.UPGRADE_DEADLINE_HEADER, None)`